src/parser/core/modules/Combos.tsx

```diff
@@ -0,0 +1,252 @@
+			.map(issue => issue.event)
+	}
+
+	protected fabricateComboEvent(event: CastEvent) {
+		const combo: ComboEvent = {
+			...event,
+			type: 'combo',
+		}
+		delete combo.timestamp // Since fabricateEvent adds that in anyway
+		this.parser.fabricateEvent(combo)
+	}
+
+	protected recordBrokenCombo(event: CastEvent, context: CastEvent[]) {
+		this.issues.push({
+			type: 'combobreak',
+			event,
+			context,
+		})
+		this.currentComboChain = []
+	}
+
+	protected recordUncomboedGcd(event: CastEvent) {
+		this.issues.push({
+			type: 'uncomboed',
+			event,
+			context: [],
+		})
+		this.currentComboChain = []
+	}
+
+	/**
+	 *
+	 * @param combo
+	 * @param event
+	 * @return true if combo, false otherwise
+	 */
+	protected checkCombo(combo: TODO /* Should be an Action type */, event: CastEvent): boolean {
+		// Not in a combo
+		if (this.lastAction == null) {
+			// Combo starter, we good
+			if (combo.start) {
+				this.fabricateComboEvent(event)
+				return true
+			}
+
+			// Combo action that isn't a starter, that's a paddlin'
+			if (combo.from) {
+				this.recordUncomboedGcd(event)
+				return false
+			}
+		}
+
+		// Incorrect combo action, that's a paddlin'
+		if (combo.from !== this.lastAction) {
+			this.recordBrokenCombo(event, this.currentComboChain)
+			return combo.start // It's a combo if the action is the start of one
+		}
+
+		// Combo continued correctly
+		this.fabricateComboEvent(event)
+		// If it's a finisher, reset the combo
+		return !combo.end
+	}
+
+	private onCast(event: CastEvent) {
+		const action = getAction(event.ability.guid) as TODO // Should be an Action type
+
+		if (!action) {
+			return
+		}
+
+		if (action.onGcd) {
+			if (event.timestamp - this.lastGcdTime > GCD_TIMEOUT_MILLIS) {
+				// If we've had enough downtime between GCDs to let the combo expire, reset the state so we don't count erroneous combo breaks
+				this.currentComboChain = []
+			}
+
+			this.lastGcdTime = event.timestamp
+		}
+
+		// If it's a combo action, run it through the combo checking logic
+		if (action.combo) {
+			const continueCombo = this.checkCombo(action.combo, event)
+			if (continueCombo) {
+				this.currentComboChain.push(event)
+			} else {
+				this.currentComboChain = []
+			}
+		}
+
+		if (action.breaksCombo && this.lastAction !== null) {
+			// Combo breaking action, that's a paddlin'
+			this.recordBrokenCombo(event, this.currentComboChain)
+		}
+	}
+
+	private onComplete() {
+		if (this.addJobSpecificSuggestions(this.comboBreakers, this.uncomboedGcds)) {
+			return
+		}
+		if (this.issues.length > 0) {
+			this.suggestions.add(new Suggestion({
+				icon: this.constructor.suggestionIcon,
+				content: <Trans id="core.combos.content">
+					Avoid misusing your combo GCDs at the wrong combo step or breaking existing combos with non-combo
+					GCDs. Breaking combos can cost you significant amounts DPS as well as important secondary effects.
+				</Trans>,
+				severity: SEVERITY.MEDIUM, // TODO
```

Comment:
this still todo?

Reply:
I could make this Tiered I guess? I just ported over the behavior that was already there

Reply:
Yeah figured it was direct port. IDM keeping straight severities if it's appropriate, just noticed the comment.